File: backend/app/models/property.py
from sqlalchemy import Boolean, Column, Integer, String, Float, ForeignKey, Enum, Text
from sqlalchemy.sql import func
from sqlalchemy.sql.sqltypes import DateTime
from sqlalchemy.orm import relationship
import enum
from ..database import Base

class PropertyType(str, enum.Enum):
    APARTMENT = "apartment"
    HOUSE = "house"
    VILLA = "villa"
    COMMERCIAL = "commercial"
    LAND = "land"

class PropertyStatus(str, enum.Enum):
    AVAILABLE = "available"
    RENTED = "rented"
    PENDING = "pending"
    MAINTENANCE = "maintenance"

class Property(Base):
    __tablename__ = "properties"

    id = Column(Integer, primary_key=True, index=True)
    title = Column(String, nullable=False)
    description = Column(Text)
    property_type = Column(Enum(PropertyType), nullable=False)
    status = Column(Enum(PropertyStatus), default=PropertyStatus.AVAILABLE)
    
    # Location
    address = Column(String, nullable=False)
    city = Column(String, nullable=False)
    district = Column(String)
    latitude = Column(Float)
    longitude = Column(Float)
    
    # Details
    bedrooms = Column(Integer)
    bathrooms = Column(Integer)
    area = Column(Float)  # in square meters
    price = Column(Float, nullable=False)  # in UGX
    currency = Column(String, default="UGX")
    
    # Features
    is_furnished = Column(Boolean, default=False)
    has_parking = Column(Boolean, default=False)
    has_security = Column(Boolean, default=False)
    has_water = Column(Boolean, default=True)
    has_electricity = Column(Boolean, default=True)
    
    # Relationships
    owner_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    owner = relationship("User", back_populates="properties")
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now()) 

File: frontend/src/types/property.ts
export type PropertyType = 'apartment' | 'house' | 'villa' | 'commercial' | 'land'
export type PropertyStatus = 'available' | 'rented' | 'pending' | 'maintenance'

export interface Property {
  id: number
  title: string
  description?: string
  property_type: PropertyType
  status: PropertyStatus
  address: string
  city: string
  district?: string
  latitude?: number
  longitude?: number
  bedrooms?: number
  bathrooms?: number
  area?: number
  price: number
  currency: string
  is_furnished: boolean
  has_parking: boolean
  has_security: boolean
  has_water: boolean
  has_electricity: boolean
  images?: string[]
  owner_id: number
  created_at: string
  updated_at?: string
}

export interface PropertySearchParams {
  property_type?: PropertyType
  min_price?: number
  max_price?: number
  city?: string
  district?: string
  bedrooms?: number
  bathrooms?: number
  is_furnished?: boolean
  has_parking?: boolean
  has_security?: boolean
  radius?: number
  latitude?: number
  longitude?: number
} 